### PR TITLE
chore(sql): move join const-filter merge to optimizer

### LINKED ISSUE #6762: AssertionError at FilteredRecordCursorFactory.java
### To reproduce

```
SELECT 
    T1.event, 
    T1.origin, 
    CASE WHEN T1.event > 50 THEN 'High' ELSE 'Low' END 
FROM telemetry T1 
INNER JOIN telemetry T2 ON T1.created < T2.created 
WHERE (cast(T1.origin as SYMBOL)) IS NULL AND NOW() = NOW() 
ORDER BY (cast(T1.origin as SYMBOL)) NOT IN ('{}') 
LIMIT NULL;
```
```
2026-02-09T13:46:03.161178Z C i.q.c.h.p.JsonQueryProcessorState [1093069874965] internal error [ex=java.lang.AssertionError
	at io.questdb.griffin.engine.table.FilteredRecordCursorFactory.<init>(FilteredRecordCursorFactory.java:44)
	at io.questdb.griffin.SqlCodeGenerator.generateJoins(SqlCodeGenerator.java:4194)
	at io.questdb.griffin.SqlCodeGenerator.generateSelect(SqlCodeGenerator.java:5344)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery0(SqlCodeGenerator.java:4839)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery(SqlCodeGenerator.java:4814)
	at io.questdb.griffin.SqlCodeGenerator.generateSubQuery(SqlCodeGenerator.java:6851)
	at io.questdb.griffin.SqlCodeGenerator.generateSelectChoose(SqlCodeGenerator.java:5358)
	at io.questdb.griffin.SqlCodeGenerator.generateSelect(SqlCodeGenerator.java:5335)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery0(SqlCodeGenerator.java:4839)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery(SqlCodeGenerator.java:4814)
	at io.questdb.griffin.SqlCodeGenerator.generateSubQuery(SqlCodeGenerator.java:6851)
	at io.questdb.griffin.SqlCodeGenerator.generateSelectVirtual(SqlCodeGenerator.java:5991)
	at io.questdb.griffin.SqlCodeGenerator.generateSelect(SqlCodeGenerator.java:5337)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery0(SqlCodeGenerator.java:4839)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery(SqlCodeGenerator.java:4814)
	at io.questdb.griffin.SqlCodeGenerator.generateSubQuery(SqlCodeGenerator.java:6851)
	at io.questdb.griffin.SqlCodeGenerator.generateSelectChoose(SqlCodeGenerator.java:5358)
	at io.questdb.griffin.SqlCodeGenerator.generateSelect(SqlCodeGenerator.java:5335)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery0(SqlCodeGenerator.java:4839)
	at io.questdb.griffin.SqlCodeGenerator.generateQuery(SqlCodeGenerator.java:4814)
	at io.questdb.griffin.SqlCodeGenerator.generate(SqlCodeGenerator.java:682)
	at io.questdb.griffin.SqlCompilerImpl.generateSelectOneShot(SqlCompilerImpl.java:4862)
	at io.questdb.griffin.SqlCompilerImpl.generateSelectWithRetries(SqlCompilerImpl.java:545)
	at io.questdb.griffin.SqlCompilerImpl.compileUsingModel(SqlCompilerImpl.java:3473)
	at io.questdb.griffin.SqlCompilerImpl.compileInner(SqlCompilerImpl.java:2830)
	at io.questdb.griffin.SqlCompilerImpl.compile(SqlCompilerImpl.java:414)
	at io.questdb.cairo.pool.SqlCompilerPool$C.compile(SqlCompilerPool.java:145)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.compileAndExecuteQuery(JsonQueryProcessor.java:493)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.execute0(JsonQueryProcessor.java:219)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.onRequestComplete(JsonQueryProcessor.java:309)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientRecv(HttpConnectionContext.java:985)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientOperation(HttpConnectionContext.java:323)
	at io.questdb.cutlass.http.HttpServer.handleClientOperation(HttpServer.java:350)
	at io.questdb.cutlass.http.HttpServer$1.lambda$$0(HttpServer.java:114)
	at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:224)
	at io.questdb.cutlass.http.HttpServer$1.run(HttpServer.java:118)
	at io.questdb.mp.Worker.run(Worker.java:152)
```

### QuestDB version:

nightly

### OS, in case of Docker specify Docker and the Host OS:

ubuntu

### File System, in case of Docker specify Host File System:

ext4

### Full Name:

Yuancheng Jiang

### Affiliation:

National University of Singapore

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [x] Yes, I have

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bluestreak01**
Follow-up to #6824 (fixes #6762).

## Summary

- `SqlOptimiser.optimiseJoins()` now merges non-compile-time-constant `constWhereClause` expressions (e.g. `NOW() = NOW()`) into the last join model's `postJoinWhereClause`, so the code generator never sees both clauses separately and cannot double-wrap `FilteredRecordCursorFactory`
- Truly constant expressions (e.g. `1 > 10`, `false`) stay as `constWhereClause` so the code generator can still optimize them away to `EmptyTableRecordCursorFactory`
- The generator-level `wrapWithFilter()` workaround and `AND_FUNCTION_FACTORY` static field from #6824 are removed; the assertion at `FilteredRecordCursorFactory:44` remains as a development-time canary
- Fixed misleading test comment on `testJoinInnerPostJoinMultipleJoinsFilter` — it tests multi-way joins with post-join filters, not the double-wrap bug

## Test plan

- `JoinTest` full suite (173 tests) passes
- `ExplainPlanTest` full suite (521 tests) passes
- `ConstantReassociationTest` (9 tests) passes
- `testJoinInnerPostJoinAndConstFilter` still covers the #6762 regression

### CODE REVIEW

**Comment by glasstiger** [comment]
@coderabbit review

**Comment by glasstiger** [comment]
@bluestreak01, comments from Claude:

-------
The core approach is sound. Merging at the AST level during optimization is architecturally cleaner than the runtime wrapWithFilter approach that combined compiled Function objects.
The optimizer already has concatFilters() for this exact pattern, and the method is used consistently elsewhere.

One concern with `isCompileTimeConstant` — mixed expressions:
  If `constWhereClause` contains both a compile-time-constant part and a non-compile-time-constant part `AND`ed together (e.g., ``(1 > 10) AND (NOW() = NOW())``), the top-level check returns false and merges the entire expression into `postJoinWhereClause`. This loses the `EmptyTableRecordCursorFactory` short-circuit that the code generator path provides.

  With the old code, the compiled AndFunction could detect that one operand is constant false and report `isConstant() == true`, letting the code generator produce `EmptyTableRecordCursorFactory`. After this PR, the same expression is compiled as a row-level filter — it still filters all rows out, but it scans the data to do so.

  This is a minor performance regression for an extremely unlikely query pattern. In practice, nobody writes `WHERE 1 > 10 AND NOW() = NOW()`. If the team cares, a possible fix would be to split the constWhereClause `AND`-tree into compile-time and non-compile-time subtrees, merging only the non-compile-time parts. But I wouldn't block on this.

-------
`isCompileTimeConstant` — Narrow but Safe

```
  private static boolean isCompileTimeConstant(ExpressionNode node) {
      if (node == null) { return true; }
      return switch (node.type) {
          case ExpressionNode.CONSTANT -> true;
          case ExpressionNode.OPERATION -> isCompileTimeConstant(node.lhs) && isCompileTimeConstant(node.rhs);
          default -> false;
      };
  }
```
  Observations:

  1. Only checks lhs/rhs, not args. OPERATION nodes with paramCount > 2 would have children in the args list, which aren't checked. This is safe in practice since boolean operations are binary, but it's worth documenting or adding an args check for robustness.
  2. Recursive without depth limit. Deeply nested constant expressions could stack-overflow. Not a realistic concern for constWhereClause which is typically shallow, but worth noting.
  3. Conservative classification. Expressions like `CAST(1 AS SHORT) > 10` contain a FUNCTION node (CAST) and would be classified as non-compile-time-constant. This is safe (they get merged) but misses an optimization. Acceptable tradeoff.
  4. null returns true. Correct for the recursive case (unary operations have rhs == null), but a standalone call with null returns true. The caller guards with `constWhere != null`, so  this isn't reachable, but the semantics are a bit surprising.

-------
Test Coverage: Adequate but Could Be Stronger

Missing:
  - No explicit test for compile-time constant preservation. A test verifying that `WHERE 1 > 10` on a join query still produces `EmptyTableRecordCursorFactory` (via explain plan) would validate that the split between compile-time and non-compile-time is working correctly.
  - No test for the edge case where constWhereClause mixes compile-time and non-compile-time parts (e.g., `WHERE (1 > 10) AND (NOW() = NOW())`). Even if the behaviour is acceptable, documenting it with a test prevents future surprises.

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 1]
```diff
@@ -3288,7 +3286,7 @@ private RecordCursorFactory generateFilter0(
                 throw e;
             }
         }
-        return wrapWithFilter(factory, filter, executionContext);
+        return new FilteredRecordCursorFactory(factory, filter);
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java]: @Coderabbit This is a false positive. The `assert !(base instanceof FilteredRecordCursorFactory)` assertion has existed since before PR #6824 — if `generateFilter0` could receive an already-filtered factory through the path described here, the assertion would have been firing in the existing test suite all along.

The scenario described (`select * from (select * from x where a > 0) where b > 0`) doesn't produce a raw `FilteredRecordCursorFactory` at the `generateFilter0` call site — subqueries go through `generateSubQuery()`, which wraps the result in a different factory type. The two wrap sites in `generateJoins()` (lines 5040 and 5103) also don't receive already-filtered factories because the optimizer now prevents double-wrapping by merging runtime const terms into `postJoinWhereClause` before the code generator runs.

The original bug (double-wrapping) is now prevented at the optimizer level, so the generator-level `wrapWithFilter()` workaround is no longer needed. The assertion stays as a canary.

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java [hunk 2]
```diff
@@ -4523,6 +4523,33 @@ private void mergeWindowSpec(WindowExpression child, WindowExpression base) {
...
+            int lastIndex = ordered.getQuick(ordered.size() - 1);
+            QueryModel lastModel = model.getJoinModels().getQuick(lastIndex);
+            lastModel.setPostJoinWhereClause(concatFilters(
+                    configuration.getCairoSqlLegacyOperatorPrecedence(),
+                    expressionNodePool,
+                    lastModel.getPostJoinWhereClause(),
+                    constWhere
+            ));
+            model.setConstWhereClause(null);
+        }
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/SqlOptimiser.java]: @Coderabbit Fixed in the second commit (3022bf61d1). `mergeConstIntoPostJoinWhereClause` now uses `extractAndTerms()` to flatten the AND-tree and classifies each conjunct independently with `isCompileTimeConstant()`. Compile-time terms stay in `constWhereClause`, runtime terms move to `postJoinWhereClause`.

Tests added for both mixed cases: `WHERE ... AND 1 > 10 AND NOW() = NOW()` (Empty table) and `WHERE ... AND 1 < 10 AND NOW() = NOW()` (rows returned, plan asserted).

`isCompileTimeConstant` also moved to the static method section per the nitpick about member ordering.

**Comment by bluestreak01** [comment]
@glasstiger thanks for the thorough review. Addressing each point:

**Mixed expressions (`(1 > 10) AND (NOW() = NOW())`)** — Fixed in the second commit (3022bf61d1). `mergeConstIntoPostJoinWhereClause` now flattens the AND-tree with `extractAndTerms()` and classifies each conjunct independently. Compile-time terms stay in `constWhereClause` (preserving `EmptyTableRecordCursorFactory` folding), runtime terms move to `postJoinWhereClause`.

**`isCompileTimeConstant` observations:**
1. *args list not checked* — correct, boolean operations are always binary (paramCount=2, lhs/rhs only). No OPERATION node uses the args list.
2. *Recursive depth* — agreed, not a practical concern. `constWhereClause` is built from `assignFilters` conjuncts, which are individual filter terms AND-ed together — the tree depth equals the number of zero-table-reference terms in the WHERE clause.
3. *CAST classified as non-compile-time* — yes, this is conservative but safe. CAST is a FUNCTION node, so `CAST(1 AS SHORT) > 10` gets merged into `postJoinWhereClause` and evaluated at runtime. Correct results, just misses the empty-table shortcut for an unrealistic query.
4. *null returns true* — only reachable in the recursive case (unary operations with null rhs). The top-level caller guards with `constWhere == null` early return.

**Test coverage** — both gaps are now covered:
- `testNoArgMixedConstAndRuntimeExprInJoin` (ExplainPlanTest) — asserts `Empty table` plan for `WHERE 1 > 10 AND NOW() = NOW()` on a join
- `testJoinInnerPostJoinAndMixedConstFilter` (JoinTest) — asserts empty result for mixed false + runtime
- `testJoinInnerPostJoinAndMixedConstTrueFilter` (JoinTest) — asserts correct result + plan for mixed true + runtime, verifying `NOW()=NOW()` appears in the post-join filter

**Review by glasstiger — DISMISSED** [review]

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 36 / 37 (97.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|34|34|100.00%|
